File: ExomineAPI/Service/FacilityMineralService.cs
using System.Collections.Generic;
using System.Linq;
using ExomineAPI.Data;
using ExomineAPI.Models;
using ExomineAPI.Models.DTOs;

namespace ExomineAPI.Services
{
    public class FacilityMineralService
    {
        public IEnumerable<FacilityMineralDTO> GetAllFacilityMinerals(
            bool expandFacility = false,
            bool expandMineral = false,
            int? facilityId = null
        )
        {
            var query = FacilityMineralData.FacilityMinerals.AsEnumerable();

            // Filter by facilityId if provided
            if (facilityId.HasValue)
            {
                query = query.Where(fm => fm.FacilityId == facilityId.Value);
            }

            return query.Select(fm =>
            {
                var facilityMineralDto = new FacilityMineralDTO
                {
                    Id = fm.Id,
                    FacilityId = fm.FacilityId,
                    MineralId = fm.MineralId,
                    Quantity = fm.Quantity,
                };

                // Expand Facility if requested
                if (expandFacility)
                {
                    var facility = FacilityData.Facilities.FirstOrDefault(f =>
                        f.Id == fm.FacilityId
                    );
                    if (facility != null)
                    {
                        facilityMineralDto.Facility = new FacilityDTO
                        {
                            Id = facility.Id,
                            Name = facility.Name,
                            Active = facility.Active,
                        };
                    }
                }

                // Expand Mineral if requested
                if (expandMineral)
                {
                    var mineral = MineralData.Minerals.FirstOrDefault(m => m.Id == fm.MineralId);
                    if (mineral != null)
                    {
                        facilityMineralDto.Mineral = new MineralDTO
                        {
                            Id = mineral.Id,
                            Name = mineral.Name,
                        };
                    }
                }

                return facilityMineralDto;
            });
        }

        public FacilityMineralDTO GetFacilityMineralById(int id)
        {
            var facilityMineral = FacilityMineralData.FacilityMinerals.FirstOrDefault(fm =>
                fm.Id == id
            );
            if (facilityMineral == null)
                return null;

            return new FacilityMineralDTO
            {
                Id = facilityMineral.Id,
                FacilityId = facilityMineral.FacilityId,
                MineralId = facilityMineral.MineralId,
                Quantity = facilityMineral.Quantity,
            };
        }

        public FacilityMineralDTO UpdateFacilityMineral(
            int id,
            FacilityMineralDTO updatedFacilityMineral
        )
        {
            var facilityMineral = FacilityMineralData.FacilityMinerals.FirstOrDefault(fm =>
                fm.Id == id
            );

            if (facilityMineral == null)
            {
                return null; // Return null if the facility mineral with the given id is not found
            }

            // Update the properties
            facilityMineral.Quantity = updatedFacilityMineral.Quantity;

            // Return the updated DTO
            return new FacilityMineralDTO
            {
                Id = facilityMineral.Id,
                FacilityId = facilityMineral.FacilityId,
                MineralId = facilityMineral.MineralId,
                Quantity = facilityMineral.Quantity,
            };
        }

        public FacilityMineralDTO CreateFacilityMineral(FacilityMineralDTO newFacilityMineral)
        {
            // Generate a new ID
            var newId = FacilityMineralData.FacilityMinerals.Max(fm => fm.Id) + 1;

            // Create the new FacilityMineral
            var facilityMineral = new FacilityMineral
            {
                Id = newId,
                FacilityId = newFacilityMineral.FacilityId,
                MineralId = newFacilityMineral.MineralId,
                Quantity = newFacilityMineral.Quantity,
            };

            // Update the database
            FacilityMineralData.FacilityMinerals.Add(facilityMineral);

            // Return the new DTO
            return new FacilityMineralDTO
            {
                Id = facilityMineral.Id,
                FacilityId = facilityMineral.FacilityId,
                MineralId = facilityMineral.MineralId,
                Quantity = facilityMineral.Quantity,
            };
        }
    }
}

File: ExomineAPI/client/exomine-client/src/services/facilityMineralService.js
import { baseURL } from "./colonyMineralService";

export const getAllFacilityMinerals = async () => {
  const response = await fetch(`${baseURL}/facilityMinerals`);
  if (!response.ok) throw new Error("Failed to fetch facility minerals.");
  return response.json();
};

export const getExpandedFacilityMinerals = async (facilityId = null) => {
  const params = new URLSearchParams({ _expand: "facility,mineral" });
  if (facilityId) params.append("facilityId", facilityId);
  const response = await fetch(
    `${baseURL}/facilityMinerals?${params.toString()}`
  );
  if (!response.ok)
    throw new Error("Failed to fetch expanded facility minerals.");
  return response.json();
};

export const getFacilityMineralById = async (id) => {
  const response = await fetch(`${baseURL}/facilityMinerals/${id}`);
  if (!response.ok)
    throw new Error(`Facility mineral with ID ${id} not found.`);
  return response.json();
};

export const createFacilityMineral = async (newFacilityMineral) => {
  const response = await fetch(`${baseURL}/facilityMinerals`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(newFacilityMineral),
  });
  if (!response.ok) throw new Error("Failed to create facility mineral.");
  return response.json();
};

export const updateFacilityMineral = async (id, updatedFacilityMineral) => {
  const response = await fetch(`${baseURL}/facilityMinerals/${id}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(updatedFacilityMineral),
  });
  if (!response.ok)
    throw new Error(`Failed to update facility mineral with ID ${id}.`);
  return response.json();
};
